mars/core/session.py

```diff
@@ -29,6 +29,40 @@
 _get_session_lock = asyncio.Lock(loop=_loop)
 
 
+def _wrap_in_thread(func: Callable):
+    """
+    Function is those wrapping async function,
+    when there is a running event loop,
+    error will raise (see GH#2108),
+    so we wrap them to run in a thread.
+    """
+
+    def inner(*args, **kwargs):
+        result = None
+        default_session_in_thread = None
+        default_session = get_default_session()
+
+        def run_in_thread():
+            nonlocal result, default_session_in_thread, default_session
+
+            if default_session:
+                # set default session in this thread
+                default_session.as_default()
+
+            result = func(*args, **kwargs)
+            default_session_in_thread = get_default_session()
+
+        t = threading.Thread(target=run_in_thread)
+        t.start()
+        t.join()
+
+        if default_session_in_thread:
+            default_session_in_thread.as_default()
+
+        return result
+    return inner
```

Comment:
One thread per call? I think we can create only one daemon thread for the event loop. Then send the async call to the event loop and get the result.

```python
import asyncio
import threading

loop = asyncio.new_event_loop()
thrd = threading.Thread(target=loop.run_forever)
thrd.daemon = True
thrd.start()


async def async_call():
    return 1


def sync_call():
    return asyncio.run_coroutine_threadsafe(async_call(), loop).result()


print(sync_call())
```

Reply:
Now I used ThreadPoolExecutor instead, check the latest commit to see if it's ok?

Reply:
There is only one thread in the `ThreadPoolExecutor` to ensure that the loop interface call will not have thread safety issues. It's OK.

Is it OK to copy the default session between main thread and worker pool thread? It seems that the thread local var `AbstractSession._default_session_local` in multiple threads hold one same session instance?

Reply:
Yeah, it should be in the same thread, due to the reason of event loop, they have to be in a different one, thus the default session should be synced in and out the thread.